server-spi-private/src/main/java/org/keycloak/events/EventListenerProvider.java

```diff
@@ -18,15 +18,52 @@
 package org.keycloak.events;
 
 import org.keycloak.events.admin.AdminEvent;
+import org.keycloak.models.KeycloakTransaction;
 import org.keycloak.provider.Provider;
 
 /**
  * @author <a href="mailto:sthorger@redhat.com">Stian Thorgersen</a>
+ *
+ * Note that invocation of any of {@code onEvent} methods doesn't mean that the event happened at the moment but that it
+ * will happen during the main transaction commit. This means the execution, corresponding to the triggered event,
+ * may end up with a failure. Therefore, each implementation of this interface should consider using an internal
+ * {@link KeycloakTransaction} for firing events only in case the main transaction ends up successfuly. This can be
```

Comment:
_Firing_ an event should be precondition for this interface's methods to be invoked. Maybe _persisting_ would express the intention? (same in the javadocs below)

Reply:
WDYT?

```Therefore, each implementation of this interface should consider persisting events and act upon them only when the main transaction was successful. This can be achieved by enlisting the internal transaction to {@link org.keycloak.models.KeycloakTransactionManager} using the {@link org.keycloak.models.KeycloakTransactionManager#enlistAfterCompletion(KeycloakTransaction)} method.```